File: leetcode/easy/144-preorder/morris.cpp
/**
 * @file morris.cpp
 * @author Vortexx2
 * @brief Problem 144 - Binary Tree Preorder Traversal
 * @date 22-07-2021
 * 
 * Runtime - 0 ms
 * Memory Usage - 8.3 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  vector<int> preorderTraversal(TreeNode *root)
  {
    vector<int> res;

    while (root)
    {
      if (root->left)
      {
        TreeNode *pred = findPredecessor(root);

        if (pred->right)
        {
          pred->right = NULL;
          root = root->right;
        }
        else
        {
          pred->right = root;
          res.push_back(root->val);
          root = root->left;
        }
      }

      else
      {
        res.push_back(root->val);
        root = root->right;
      }
    }

    return res;
  }

  TreeNode *findPredecessor(TreeNode *node)
  {
    TreeNode *curr = node->left;
    while (curr->right && curr->right != node)
      curr = curr->right;

    return curr;
  }
};

File: leetcode/easy/144-preorder/morris.py
"""
  @author Vortexx2
  Problem 144 - Binary Tree Preorder Traversal

  Runtime - 24 ms
  Memory Usage - 14.2 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def preorderTraversal(self, root: TreeNode) -> List[int]:
    res = []

    while root:
      if (root.left):
        predec = self.find_predecessor(root)

        if (predec.right):
          predec.right = None
          root = root.right

        else:
          predec.right = root
          res.append(root.val)
          root = root.left

      else:
        res.append(root.val)
        root = root.right

    return res

  def find_predecessor(self, node: TreeNode) -> TreeNode:
    curr = node.left

    while curr.right and (curr.right != node):
      curr = curr.right

    return curr
